src/main/java/com/celements/store/DocumentCacheStore.java

```diff
@@ -0,0 +1,977 @@
+      } else if (beforeState == 0) {
+        invalidState = InvalidateState.LOADING_CANCELED;
+      } else {
+        invalidState = InvalidateState.LOADING_CANCEL_FAILED;
+      }
+      boolean succInvalidated = beforeState <= 0;
+      LOGGER_DL.debug("invalidated cache during loading document. {}", succInvalidated);
+      return invalidState;
+    }
+
+    /**
+     * IMPORTANT: do not change anything on the synchronization of this method.
+     * It is a very delicate case and very likely memory visibility breaks in less than 1 out of
+     * 100'000 document loads. Thus it is difficult to test for correctness.
+     */
+    private XWikiDocument loadDocument(String key, XWikiDocument doc, XWikiContext context)
+        throws XWikiException {
+      checkArgument(key);
+      if (loadedDoc == null) {
+        synchronized (this) {
+          if (loadedDoc == null) {
+            // if a thread is just between the document cache miss and getting the documentLoader
+            // when the documentLoader removes itself from the map, then a new documentLoader is
+            // generated. Therefore we double check here that still no document is in cache.
+            loadedDoc = getCache().get(key);
+            if (loadedDoc == null) {
+              XWikiDocument newDoc = null;
+              do {
+                // use a further synchronized method call to prevent an unsafe publication of the
+                // new document over the cache
+                if ((loadingState.getAndSet(_DOCSTATE_LOADING) < _DOCSTATE_LOADING)
+                    && (newDoc != null)) {
+                  LOGGER_DL.info("DocumentLoader-{}: invalidated docloader '{}' reloading",
+                      Thread.currentThread().getId(), key);
+                }
+                newDoc = new DocumentBuilder().buildDocument(key, doc, context);
+              } while (!loadingState.compareAndSet(_DOCSTATE_LOADING, _DOCSTATE_FINISHED));
+              LOGGER_DL.info("DocumentLoader-{}: put doc '{}' in cache",
+                  Thread.currentThread().getId(), key);
+              getCache().set(key, newDoc);
+              getPageExistCache().set(key, Boolean.valueOf(!newDoc.isNew()));
+              loadedDoc = newDoc;
+            } else {
+              LOGGER_DL.info("DocumentLoader-{}: found in cache skip loding for '{}'",
+                  Thread.currentThread().getId(), key);
+            }
+            documentLoaderMap.remove(key);
+          }
+        }
+      }
+      return loadedDoc;
+    }
+
+    private void checkArgument(String key) {
+      if (!this.key.equals(key)) {
+        throw new RuntimeException(
+            "DocumentLoader illegally used with a different key (registered key:" + this.key
+                + ", loading doc key: " + key + ").");
+      }
+    }
+
+    private class DocumentBuilder {
+
+      private synchronized XWikiDocument buildDocument(String key, XWikiDocument doc,
```

Comment:
this method needn't be synchronized, it is only used in _DocumentLoader.loadDocument_ where a new instance is created for every method call anyway